design-proposals/003-credentials.md

```diff
@@ -0,0 +1,158 @@
+
+## Discussion 
+
+The idea of credential provider is to encapsulate managing specifics of 3rd party SaaS credentials.
+Currently credentials are represented by connection properties that are marked as secret. This proposes an abstraction over existing connection properties with the ability to implement specific functionality for getting credentials such as OAuth tokens from 3rd party SaaS solutions. As such existing connection properties and their persistance as k8s secrets would remain the same, the UI would also have the choice of either managing the connection properties in the same way as it is currently done or can call into credentials provider implementation, specific to each SaaS solution, that would *apply* credentials to connection properties on the backend.
+Each 3rd party SaaS can have slightly different OAuth flow which mandates the need for credential provider, a plug in, possibly shipped as a separate artifact and discovered at runtime, mounted at `/api/{version}/credentials/{connector}` to facilitate the OAuth flow and manage (i.e. *apply*) the resulting credentials.
+Considering that OAuth requires per-user and per-application secrets, there is a need for system (Syndesis owned) and user (citizen user owned) credentials. Melding of those would be also responsibility of credential provider.
+Management of user and system credentials is also performed by credential provider, for instance initial setup of system credentials or refresh of user credentials.
+Credential provider consists of:
+ * **handler** - to implement OAuth flow (REST)
+ * **applicator** - to apply credentials to connection properties
+ * **deployer** - to augment k8s deployment descriptor
+ * **manager** - to manage credentials
+
+## Domain
+
+There are two types of domain objects, differentiated by security: public data for UI and sensitive secrets for the backend.
+
+Credentials need to be combined or linked together with other credentialsm, this has merits when considering that OAuth credentials are partialy owned by the user and partialy owned by the system. For example Syndesis owns non-public credentials (OAuth Client Secret, X.509 certificate and private key) and semi-public (OAuth Client ID), that need to be used during acquisition and in a running integration together with private credentials associated with the user (username, OAuth authorization token, OAuth refresh token). This makes for two credentials with two different security and lifecycle requirements.
+
+NOTE: Salesforce supports using x509 certificates for OAuth flow, and that should be prefered as the refreshing OAuth token can be performed within certificate's expiry i.e. not depending on the refresh token.
+
+### Credential view model
+
+Each credential should have a name so that the user can identify it, type: this can be username-password, x509 or OAuth tokens, internal identifier and connector label so it can be determined to what connection it can be applied to.
```

Comment:
I think I need provide more detail about this, I would like to have list of credentials that the client has provided Syndesis, and if they are solely defined within connection properties we can't have metadata such as name or description. Right now I'm thinking that these can be stored in k8s secrets as annotations, but I'm not sure if the link between connection property and secret can be followed.